enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -348,78 +377,58 @@ impl<'a> Handler<'a> {
+        // Update pointers from iovecs to buffers in untrusted memory.
+        let mut size = 0;
+        for (t, u) in tiovec.iter().zip(uiovec.iter_mut()) {
+            size += t.iov_len;
+            u.iov_base = c.alloc(t.iov_len).unwrap().as_mut_ptr();
         }
 
-        // Set pointers in unencrypted iovec slice to use the rest of the allocated memory.
-        // The offset is into the buffer area allocated immediately after the iovec struct
-        // array, measured in bytes.
-        let mut offset = 0;
-        for (t, mut u) in trusted.iter_mut().zip(untrusted.iter_mut()) {
-            u.iov_base = ubuffer[offset..].as_mut_ptr() as *mut _;
-            u.iov_len = t.iov_len;
-            offset += t.iov_len;
-        }
+        let ptr: Register<usize> = uiovec.as_ptr().into();
+        let len: Register<usize> = uiovec.len().into();
 
         // Do the syscall; replace encrypted memory with unencrypted memory.
-        let ret = unsafe {
-            self.syscall(
-                libc::SYS_readv as u64,
-                fd,
-                untrusted.as_ptr() as _,
-                untrusted.len() as u64,
-                0,
-                0,
-                0,
-            )
+        let res = unsafe {
+            self.proxy(Request::new(
+                libc::SYS_readv as usize,
+                &[fd.into(), ptr, len],
+            ))
         };
 
-        // Copy the unencrypted input into encrypted memory.
-        if ret <= ERRNO_BASE {
-            if ret > size as u64 {
-                self.attacked();
-            }
-
-            let mut offset = 0;
-            for (t, u) in trusted.iter_mut().zip(untrusted.iter_mut()) {
-                let us = &mut ubuffer[offset..][..t.iov_len];
-                offset += t.iov_len;
-
-                if u.iov_base != us.as_mut_ptr() as *mut _ || u.iov_len != t.iov_len {
+        match res {
+            Ok(res) => {
+                if res[0].raw() as u64 > size as u64 {
                     self.attacked();
                 }
 
-                let ts = unsafe { from_raw_parts_mut(t.iov_base as *mut u8, t.iov_len) };
-                ts.copy_from_slice(us);
+                // Allocate the correct number of iovecs in untrusted memory.
+                // We reallocate again here to keep the Rust compiler happy.
+                let mut c = Cursor::new(&mut self.block.buf);
+                let uiovec: &mut [libc::iovec] = c.alloc(tiovec.len()).unwrap();
+
+                for (t, u) in tiovec.iter_mut().zip(uiovec.iter_mut()) {
+                    u.iov_base = c.alloc(t.iov_len).unwrap().as_mut_ptr();
+                    // Grab data from untrusted buffers and copy them to trusted buffers
+                    let ubuf = unsafe { from_raw_parts(u.iov_base as *mut u8, u.iov_len) };
+                    let tbuf = unsafe { from_raw_parts_mut(t.iov_base as *mut u8, t.iov_len) };
+                    tbuf.copy_from_slice(ubuf);
+                }
```

Comment:
We probably need to be more defensive in this section. `res[0]` returns the number of bytes read. We should copy no more than this number of bytes into the trusted memory.

Reply:
I've added a check for this.